### ISSUE TITLE
QGIS Vanilla Install

### ISSUE THREAD

**Issue by MattReimer** [Riverscapes/RiverscapesToolbar#9]
@volkcj mentioned this and I agree: Installing OSGEO4W64 properly is an involved process. Too involved for mass consumption. Would be nice to be able to run our tools on a vanilla install of QGIS but we need to investigate if this is even possible.

**Questions that need answering:**

* How do we install pip dependencies?

**Comment by jesselangdon** [comment]
I agree with @MattReimer and @volkcj ... I've been wondering if there isn't a way to automatically have a tool (i.e. the Riverscapes toolbar, network profiler, pyGNAT, etc.) use easy_install.py to try installing pip if it isn't present in the user's local environment, then using pip to install required libraries?

**Comment by philipbaileynar** [comment]
This knowledge is out there. Are there people at QGIS or other foundations that we can pay for 1-5 hrs of consulting for some of this strategic knowledge. Cheaper than hours of research by us alone.

**Comment by volkcj** [comment]
@jesselangdon  are you still tied in with some of the Puget Sound Open Source developer groups?  If so, it's worth pinging them.  

We can probably squeak by for August if we had to, but we should have a plan scoped before then and in place by this fall, but the current process isn't feasible for the general user community.

**Comment by jesselangdon** [comment]
I'm not really tied in with them, and they tend to be more focused on web mapping anyway.

Here are a couple of interesting potential solutions regarding bootstrapping pip:

- https://stackoverflow.com/a/25643988/1618640
- **ensurepip** - https://docs.python.org/2/library/ensurepip.html

**Comment by MattReimer** [comment]
@philipbaileynar I just did a QGIS install and was surprised to find that both `pip` AND `easy_install` come with it. (I swear this wasn't true the last time I tried it). 

A downside is that it doesn't come with `git` so our method of installing from a git repo `pip install git+https://github.com/pypa/pip.git` won't work but maybe we can either get our pip packages accepted by PyPi [(instructions don't look that hard)](http://peterdowns.com/posts/first-time-with-pypi.html) or host it ourselves.

@jesselangdon as for bootstrapping `pip`, I think there are some problems. The chief among them is that python environments can be carefully created and maintained. This isn't precisely true for windows but on OSX or linux when you install using pip you're messing with the user's whole system, not just their QGIS install. 

From that first link @jesselangdon  sent: ***"Don't touch user's installation."***

Good advice.  Take it from a linux user. If you mess with that people will be PISSED!

I do like this solution though:

```python
try:
    import foo
except ImportError:
    sys.exit("""You need foo! install it from http://pypi.python.org/pypi/foo or run pip install foo.""")
```

A third option is that we look at we look at bundling `.egg` files with our plugin. Since none of our code needs to be compiled I think we might be able to bundle it safely but there are namespace issues we might need to deal with. 

[Someone here claims they have a solution for that](http://lists.osgeo.org/pipermail/qgis-developer/2016-January/040985.html) but I'm a bit skeptical.

**Comment by MattReimer** [comment]
Followup to previous comment:

Boundless QGIS Plugins use a scripting [package called "Paver"](https://github.com/boundlessgeo/qgis-webappbuilder-plugin/blob/master/pavement.py) to automate packaging dependencies with their plugins. 

Sarcastic Sidenote: ***"Yay! Another scripting framework to learn"***

It should be noted that all the dependencies they use they write themselves so there may be legal considerations in packaging a 3rd party app like "Boto3"

**Comment by philipbaileynar** [comment]
I can confirm that easy_install came with the latest clean OSGeo install that I just put on my computer.

**Comment by MattReimer** [comment]
Nope. You got `easy_install` when you selected to install "setuptools". It doesn't come default for the OSGeo install. 

Let's separate out some common cases:

1. Standalone QGis install. Pretty self-explanatory
    * Comes with `easy_install` and `pip` but you need to install boto3 etc yourself.
2. OSGeo Default install. 
    * I don't know what this comes with but I assume it's pretty similar to the feature spec of the standalone install. More research needed.
3. OSGeo Advanced install
    * This is what most of our developers, @philipbaileynar, @KellyMWhitehead, @jesselangdon etc should be using. 

This topic gets super confusing because we need to keep OSGeo and QGIS standalone separate in our heads and which packages come installed standard with each. 

**As a general rule:** Only developers who need to debug and deploy plugins should use OSGeo. Everyone else and their aunts should use standalone.
